[psc-ide] Polling option for psc-ide-server

kRITZCREEK opened the pull request:
This is investigative work for #2209. It forces the use of polling for the filewatcher.

@garyb or @natefaubion could you give this branch a try? Right now this branch ignores the `--force-polling` flags and always polls, but if it works out I'll add the option and we can add a toggle to the editor plugins.

garyb commented:
Thanks - I'll give this a try soon, it'll probably be once we're done updating SlamData for 0.10, since it'll be easier to replicate there :wink:

kRITZCREEK commented:
Oh right... I could make a branch and backport it to 0.8.5, unless the slamdata update is coming soon^^

garyb commented:
We're not that out of date! Currently on 0.9.3 - but yep, the update is underway: https://github.com/slamdata/slamdata/issues/1185

garyb commented:
Giving this a try, finally!

garyb commented:
It's not perfect - I have a case here where I've changed the definition of a type in one file and it's not being picked up in another - but it's _way_ better (it's actually usable again :smile:). I have a feeling the problem I'm describing there happens some other way though, as I've had it happen on other projects where `psc-ide-server` isn't suffering from this particular problem.

kRITZCREEK commented:
That sounds fantastic! In that case I'll try the polling option a bit while playing around in pscid and if it doesn't degrade my experience I'll set is as the default and otherwise make it an option.

garyb commented:
I probably should have mentioned in the above, that was on Windows at least. I'm unable to test on OSX since I can't compile it currently (#2426).

kRITZCREEK commented:
After more testing with the current master and this branch @natefaubion and @garyb report that the file watcher works fine as-is on OSX and needs the polling option on Windows, so I've made polling the default on Windows and added a flag `--polling` that toggles the polling behaviour to the `psc-ide-server` executable.

I don't have a proper Windows Haskell development environment atm, so it would be nice if someone could check that the flag is actually reversed on Windows. I added logging of the parsed options when the server is started with `--debug`.

kRITZCREEK commented:
I gave this a spin on windows and made a minor modification, but it seems to detect the OS properly now. Ready to go if CI passes.

paf31 commented:
👍 Great!
